contracts/contracts/chains/ethereum/EthereumMessengers.sol

```diff
@@ -0,0 +1,106 @@
+    function callAllowed(address, address) external pure returns (bool) {
+        // The call must be delivered by the Resolver
+        // Since we are on the same chain as the Resolver
+        // we should never see an execution of this function.
+        revert("Unexpected state.");
+    }
+
+    // Simply forward the message to the target on L1
+    function sendMessage(address target, bytes calldata message)
+        external
+        restricted(block.chainid)
+    {
+        target.call(message);
+    }
+}
+
+// In order to keep naming consistent between our rollup specific contracts
+// we are using `EthereumL2Messenger` here as a name even if this contract
+// is supposed to be deployed on L1.
+contract EthereumL2Messenger is IMessenger, Ownable, RestrictedCalls {
+    enum MessageStatus {
+        UNDEFINED,
+        DEFINED,
+        RELAYED
+    }
+
+    Resolver public immutable resolver;
+
+    /// Maps message hashes with their status.
+    mapping(bytes32 => MessageStatus) public messageHashes;
+
+    constructor(address resolver_) {
+        resolver = Resolver(resolver_);
+    }
+
+    function callAllowed(address, address) external pure returns (bool) {
+        // Since we are on the same chain as the FillManager & RequestManager
+        // we should never see an execution of this function.
+        revert("Unexpected state.");
+    }
+
+    // Store message instead of forwarding it.
+    // Can be forwarded, if ever needed, by calling `relayMessage`.
+    //
+    // .. seealso:: :sol:func:`relayMessage`
+    function sendMessage(address, bytes calldata message)
+        external
+        restricted(block.chainid)
+    {
+        bytes32 messageHash = keccak256(message);
+
+        require(
+            messageHashes[messageHash] != MessageStatus.DEFINED,
+            "Message is already registered."
+        );
+        require(
+            messageHashes[messageHash] != MessageStatus.RELAYED,
+            "Message has already been relayed."
+        );
+
+        messageHashes[messageHash] = MessageStatus.DEFINED;
+    }
+
+    function relayMessage(
+        bytes32 requestId,
+        bytes32 fillId,
+        uint256 sourceChainId,
+        address filler
+    ) external {
+        bytes32 messageHash = keccak256(
+            abi.encodeCall(
+                Resolver.resolve,
+                (requestId, fillId, block.chainid, sourceChainId, filler)
+            )
+        );
+
+        require(
+            messageHashes[messageHash] != MessageStatus.UNDEFINED,
+            "Message not yet registered."
+        );
+        require(
+            messageHashes[messageHash] != MessageStatus.RELAYED,
+            "Message has already been relayed."
+        );
+
+        messageHashes[messageHash] = MessageStatus.RELAYED;
+
+        resolver.resolve(
+            requestId,
+            fillId,
+            block.chainid,
+            sourceChainId,
+            filler
+        );
+    }
```

Comment:
Why not do

```
address(resolver).call(data)
```
where `data` is the result of `abi.encodeCall` above, which you have to do anyway? We would save unnecessary stack operations that way and make it more robust against potential breakages where we change the arguments in one place, but not the other.

Of course, similar to a previous comment, `call` requires proper return value checks.

Reply:
Good idea 👍